Fix seccomp output in `docker info`

yongtang opened the pull request:
This fix tries to address the issue raised in #24374 where `docker info` outputs seccomp support in Ubuntu 14.04 but the seccomp wass not actually supported.

The issue is that in the current docker implementation, seccomp support is only checked against the kernel by inspect `CONFIG_SECCOMP` and `CONFIG_SECCOMP_FILTER`. However, seccomp might not be enabled when building docker (through golang build flag).

This fix adds a `supportsSeccomp` boolean variable. The `supportsSeccomp` is only set to true when seccomp is enabled while building docker.

This fix fixes #24374.

Signed-off-by: Yong Tang yong.tang.github@outlook.com

yongtang committed:
Fix seccomp output in `docker info`

This fix tries to address the issue raised in #24374 where
`docker info` outputs seccomp support in Ubuntu 14.04 but
the seccomp wass not actually supported.

The issue is that in the current docker implementation, seccomp
support is only checked against the kernel by inspect CONFIG_SECCOMP
and CONFIG_SECCOMP_FILTER. However, seccomp might not be enabled
when building docker (through golang build flag).

This fix adds a supportSeccomp boolean variable. The supportSeccomp
is only set to true when seccomp is enabled when building docker.

This fix fixes #24374.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

thaJeztah commented:
ping @amitkris do we need something for Solaris, or is that using the `_linux.go` file?

amitkris commented:
@thaJeztah : Solaris will be using  _disabled.go atleast initially, so we're covered. Thank you for reaching out! :)

thaJeztah commented:
@amitkris cool, thank you!


@yongtang doesn't windows use `_disabled.go` as well?

yongtang commented:
Thanks @thaJeztah @amitkris for the review.

@thaJeztah I had some confusion about the build tag too. Before my commit, the build tag was 
`// +build !seccomp,!windows` and  `seccomp,linux`, and it caused build error on `windowsRS1` when I added `supportsSeccomp`.

That was why I added the _windows.go

Let me take a deeper look into it and see my understanding is correct or not.

thaJeztah commented:
Thanks! Perhaps it's okay, was just wondering if we could just remove the `!windows` from the `_unsupported.go` file, but maybe not

yongtang commented:
@thaJeztah I take a second look. Both _disabled.go and _linux.go uses `SeccompProfile` which only exist in `linux OR freebsd` (see container_unix.go).

Based on it I think the following would cover all cases (*comma* as *AND*) and easier to maintain:

_linux.go:
`// +build linux,seccomp`

_disabled.go:
`// +build linux,!seccomp`

_unsupported.go: 
`// +build !linux`

I updated the PR based on above. Please let me know if there are any issues.

thaJeztah commented:
nice! LGTM (if CI is green 😇 )

justincormack commented:
LGTM, thanks for fixing.
